spec/models/screened_ip_address_spec.rb

```diff
@@ -320,4 +320,63 @@ def test_bad_value(arg)
       end
     end
   end
+
+  describe '#roll_up' do
+    it 'rolls up IPv4 addresses' do
+      # roll up to /24 requires 256 * 1% ~ 3 IPs
+      # roll up to /22 requires 1024 * 1% ~ 10 IPs
+      SiteSetting.min_ban_entries_for_roll_up = 1
+
+      Fabricate(:screened_ip_address, ip_address: "1.1.1.1")
+      expect { ScreenedIpAddress.roll_up }.not_to change { ScreenedIpAddress.count }
+
+      Fabricate(:screened_ip_address, ip_address: "1.1.1.2")
+      expect { ScreenedIpAddress.roll_up }.not_to change { ScreenedIpAddress.count }
+
+      Fabricate(:screened_ip_address, ip_address: "1.1.1.3")
+      expect { ScreenedIpAddress.roll_up }.to change { ScreenedIpAddress.count }.by(-2)
+      expect(ScreenedIpAddress.pluck(:ip_address)).to include("1.1.1.0/24")
+      expect(ScreenedIpAddress.pluck(:ip_address)).not_to include("1.1.1.1", "1.1.1.2", "1.1.1.3")
+
+      # roll up is stable
+      expect { ScreenedIpAddress.roll_up }.to change { ScreenedIpAddress.count }.by(0)
+      expect(ScreenedIpAddress.pluck(:ip_address)).to include("1.1.1.0/24")
+      expect(ScreenedIpAddress.pluck(:ip_address)).not_to include("1.1.1.1", "1.1.1.2", "1.1.1.3")
+    end
+
+    it 'rolls up IPv6 addresses' do
+      # roll up to /64 requires 256 * 1% ~ 3 IPs
```

Comment:
My initial thinking on this was that we don't want to be blocking anything smaller (longer prefix) than a /64. A /64 is a single network and getting additional addresses on that is trivial, so we may as well start there.

If we *don't* do that, then I think having a separate algorithm for that initial rollup to /64 makes sense. The default value of 5 would mean *15* IP addresses have to be blocked before it's rolled up to a /64. I think that would be a lot of annoyance to the site's admins before the rollup kicks in.

Just realised this significantly changes the behaviour of `SiteSetting.min_ban_entries_for_roll_up`, by the way. That'll need to be addressed by a change to the setting text? Or `SiteSetting.min_ban_entries_for_roll_up_percent`?

Beyond that, maybe we want to *only* roll up from entries of the next size down?